src/mococrw/ecies.h

```diff
@@ -0,0 +1,363 @@
+     */
+    void setEphemeralKey(const AsymmetricPublicKey& ephKey);
+
+    /**
+     * @brief Set the authentication tag
+     *
+     * As the last step during encryption the authentication tag is calculated. Set this tag to verify the authenticity
+     * of the received message.
+     * @param tag The authentication tag
+     */
+    void setMAC(const std::vector<uint8_t>& tag);
+
+private:
+
+    ECIESDecryptionCtx(const AsymmetricPrivateKey& privKey,
+                       const std::shared_ptr<mococrw::KeyDerivationFunction> kdf,
+                       const size_t macKeySize,
+                       std::function<std::unique_ptr<MessageAuthenticationCode> (const std::vector<uint8_t> &)>,
+                       const size_t symmetricCipherKeySize,
+                       std::function<std::unique_ptr<SymmetricCipherI>(const std::vector<uint8_t>&,
+                                                                       const SymmetricCipherOperation&)>,
+                       const std::vector<uint8_t>& S1,
+                       const std::vector<uint8_t>& S2);
+
+    friend ECIESDecryptionCtxBuilder;
+
+    /**
+     * Internal class for applying the PIMPL design pattern
+     * (to hide the details of storing the padding objects from the client)
+     */
+    class Impl;
+
+    /**
+     * Pointer for PIMPL design pattern
+     */
+    std::unique_ptr<Impl> _impl;
+};
+
+/**
+ * @brief The ECIESCtxBuilder class is an pseudo abstract class for providing common functions to the Encryption and
+ * Decryption context builders.
+ */
+class ECIESCtxBuilder
+{
+public:
+    /**
+     * @brief Sets the key derivation function used for ECIES (Default: mococrw::X963KDF with SHA512) (required)
+     * @param kdf The actual object used in ECIES
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setKDF(std::shared_ptr<mococrw::KeyDerivationFunction> kdf);
+
+    /**
+     * @brief Sets the required key size for MAC ciphers (required)
+     * @param keySize The key size in bytes
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMacCipherKeySize(size_t keySize);
+
+    /**
+     * @brief Sets a function for creating the MAC object inside ECIES (required)
+     *
+     * As the key for MAC is generated during ECIES key generation, the MAC object has to be instantiated after that.
+     * This function is a dependency injection.
+     * (e.g. for HMAC-SHA512:
+     \code{.cpp}
+     auto macFunc = [](const std::vector<uint8_t> &key) {
+        std::unique_ptr<MessageAuthenticationCode>mac_ptr(new mococrw::HMAC(openssl::DigestTypes::SHA512, key, 64));
+        return mac_ptr;
+     };
+     \endcode
+     * )
+     * @param func A lambda expression which returns an object implementing the MessageAuthenticationCode interface
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMacFactoryFunction(std::function<std::unique_ptr<MessageAuthenticationCode>(
```

Comment:
Does the IEEE standard specify a default? If so, it would be nice if users didn't have to write this piece of boilerplate.

Reply:
Unfortunately the standard doesn't define any default.